plugins/sword-bible-engine-plugin/src/OsisParser.ts

```diff
@@ -0,0 +1,443 @@
+                context.paragraph = undefined;
+            }
+            break;
+        }
+        case OsisXmlTag.DIVINE_NAME:
+            context.divineNameNode = undefined;
+    }
+    context.currentNode = undefined;
+}
+
+function getPhrase(
+    content: string,
+    context: ParserContext
+): IBibleContentPhraseForInput {
+    const strongs = getStrongsNumbers(context);
+    const phrase: IBibleContentPhraseForInput = {
+        type: 'phrase',
+        content,
+        versionChapterNum: context.chapterNum,
+        versionVerseNum: context.verseNum
+    };
+    if (strongs && strongs.length) {
+        phrase['strongs'] = strongs;
+    }
+    if (context.crossRefs.length) {
+        phrase['crossReferences'] = context.crossRefs;
+        context.crossRefs = [];
+    }
+    if (context.notes.length) {
+        phrase['notes'] = context.notes;
+        context.notes = [];
+    }
+    return phrase;
+}
+
+function getCrossReference(context: ParserContext): IBibleCrossReference {
+    const key = String(context.currentNoteNode!.attributes.n);
+    const { osisRef } = context.currentCrossRefNode!.attributes;
+    let {
+        bookOsisId,
+        versionChapterNum,
+        versionVerseNum
+    } = getOsisReferenceEntities(osisRef);
+    bookOsisId = String(bookOsisId);
+    const crossRef: IBibleCrossReference = {
+        key,
+        range: {
+            bookOsisId,
+            versionChapterNum,
+            versionVerseNum
+        }
+    };
+    return crossRef;
+}
+
+function getOsisReferenceEntities(osisRef: string): IBibleReferenceRange {
+    const firstVerse = osisRef.split('-')[0].split('.');
+    const bookOsisId = firstVerse[0];
+    const versionChapterNum = Number(firstVerse[1]);
+    const versionVerseNum = Number(firstVerse[2]);
+    const range: IBibleReferenceRange = {
+        bookOsisId,
+        versionChapterNum,
+        versionVerseNum
+    };
+    const hasMultipleVerses = osisRef.split('-').length === 2;
+    if (hasMultipleVerses) {
+        const secondVerse = osisRef.split('-')[1];
+        range.versionVerseEndNum = Number(secondVerse.split('.')[2]);
+    }
+    return range;
+}
+
+function getFootnote(context: ParserContext): IBibleNote {
+    const key = context.currentNoteNode!.attributes.n!;
+    const type: string = context.currentNoteNode!.attributes.type!;
+    const content: DocumentRoot = {
+        type: 'root',
+        contents: [
+            {
+                type: 'phrase',
+                content: context.noteText
+            }
+        ]
+    };
+    return {
+        type,
+        key,
+        content
+    };
+}
+
+function isPunctuation(str: string): boolean {
+    return str.length === 1 && !!str.match(/^[.,:;!?]/);
+}
+
+function getNewParagraph(): IBibleContentGroupForInput<'paragraph'> {
+    return {
+        type: 'group',
+        groupType: 'paragraph',
+        contents: []
+    };
+}
+
+function getIndentGroup(
```

Comment:
i don't know if you fixed this already, but maybe this method is responsible that every phrase has an indent modifier (in the bible.db you sent me)? 

concerning `Indentation.LARGE`: is this about an identation that is actually larger than others or is this rather indented content within indented content? If the former we should maybe introduce a modifier for this groupType ('small' | 'large'), if the latter, the indent group should be added to the parent one and not create two indent groups on its own.

Reply:
`Indentation.LARGE` actually does mean indent within an indent, the amount of spacing that makes up a double indent is just double the amount of space of a single indent. so it should be fine. 

You're right though about it erroneously modifying every phrase, need to fix that

Reply:
correct me if i'm wrong: but wouldn't an indent within an indent look something like this:
```
text
    indent 1
        indent 2
```
so `indent 2` is within a group that is indented already (and by that inheriting the first indent). wouldn't your code cause it to have 3 indents then?

Reply:
Oh I see what you mean. I'll only have one indentation amount, so `indent 2` will only be indented an additional amount further, which is same.

Reply:
not sure how to interpret your comment - so are you saying that there are no indents within indents in your source file (but different indent widths) or that all indents have the same width (but are relative to parent indents and thus the parser shouldn't produce a double-indent)?

maybe part of the confusion is that you are not supposed to set an indent level or size in the input but that BibleEngine infers that from the document hierarchy?